src/main/cpp/src/slice.cu

```diff
@@ -0,0 +1,358 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "slice.hpp"
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/copy.hpp>
+#include <cudf/detail/gather.hpp>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/detail/sizes_to_offsets_iterator.cuh>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/lists/filling.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/logical.h>
+
+using namespace cudf;
+
+namespace spark_rapids_jni {
+
+namespace detail {
+
+namespace {
+
+void assert_start_is_not_zero(column_device_view const& start, rmm::cuda_stream_view stream)
+{
+  bool start_valid = thrust::all_of(rmm::exec_policy(stream),
+                                    thrust::make_counting_iterator(0),
+                                    thrust::make_counting_iterator(start.size()),
+                                    [start] __device__(size_type index) {
+                                      if (start.is_null(index)) return true;
+                                      return start.element<int32_t>(index) != 0;
+                                    });
+  CUDF_EXPECTS(start_valid, "Invalid start value: start must not be 0");
+}
+
+void assert_length_is_not_negative(column_device_view const& length, rmm::cuda_stream_view stream)
+{
+  bool length_valid = thrust::all_of(rmm::exec_policy(stream),
+                                     thrust::make_counting_iterator(0),
+                                     thrust::make_counting_iterator(length.size()),
+                                     [length] __device__(size_type index) {
+                                       if (length.is_null(index)) return true;
+                                       return length.element<int32_t>(index) >= 0;
+                                     });
+  CUDF_EXPECTS(length_valid, "Invalid length value: length must be >= 0");
+}
+
+struct int_iterator_from_scalar {
+  int32_t value;
+  int_iterator_from_scalar(int32_t v) : value(v) {}
+  __device__ bool is_null(cudf::thread_index_type const index) const { return false; }
+  __device__ int32_t operator()(cudf::thread_index_type const index) const
+  {
+    // ignore index, always return the same value
+    return value;
+  }
+};
+
+struct int_iterator_from_column {
+  column_device_view cdv;
+  int_iterator_from_column(column_device_view cdv) : cdv(cdv) {}
+  __device__ bool is_null(cudf::thread_index_type const index) const { return cdv.is_null(index); }
+  __device__ int32_t operator()(cudf::thread_index_type const index) const
+  {
+    return cdv.element<int32_t>(index);
+  }
+};
+
+template <typename SRART_ITERATOR, typename LENGTH_ITERATOR>
+CUDF_KERNEL void compute_starts_and_sizes_kernel(size_type const* offsets_of_input_lists,
```

Comment:
Better to have comments to describe its functionality and the input and output for easier maintenance in the future.